src/Lean/Elab/Tactic/BvDecide/BitBlast/BVExpr/BitBlast/Impl/ShiftLeft.lean

```diff
@@ -0,0 +1,218 @@
+
+namespace BVExpr
+namespace bitblast
+
+variable [Hashable α] [DecidableEq α]
+
+def blastShiftLeftConst (aig : AIG α) (target : AIG.ShiftTarget aig w) :
+    AIG.RefVecEntry α w :=
+  let ⟨input, distance⟩ := target
+  go aig input distance 0 (by omega) .empty
+where
+  go (aig : AIG α) (input : AIG.RefVec aig w) (distance : Nat) (curr : Nat) (hcurr : curr ≤ w)
+      (s : AIG.RefVec aig curr) :
+      AIG.RefVecEntry α w :=
+  if hidx : curr < w then
+    if hdist : curr < distance then
+      let res := aig.mkConstCached false
+      let aig := res.aig
+      let zeroRef := res.ref
+      have hfinal := AIG.LawfulOperator.le_size (f := AIG.mkConstCached) ..
+      let s := s.cast hfinal
+      let input := input.cast hfinal
+      let s := s.push zeroRef
+      go aig input distance (curr + 1) (by omega) s
+    else
+      let s := s.push (input.get (curr - distance) (by omega))
+      go aig input distance (curr + 1) (by omega) s
+  else
+    have hcurr : curr = w := by omega
+    ⟨aig, hcurr ▸ s⟩
+  termination_by w - curr
+
+theorem blastShiftLeftConst.go_le_size (aig : AIG α) (distance : Nat) (input : AIG.RefVec aig w)
+    (curr : Nat) (hcurr : curr ≤ w) (s : AIG.RefVec aig curr) :
+    aig.decls.size ≤ (go aig input distance curr hcurr s).aig.decls.size := by
+  unfold go
+  split
+  . dsimp only
+    split
+    . refine Nat.le_trans ?_ (by apply go_le_size)
+      apply AIG.LawfulOperator.le_size
+    . refine Nat.le_trans ?_ (by apply go_le_size)
+      omega
+  . simp
+termination_by w - curr
+
+theorem blastShiftLeftConst.go_decl_eq (aig : AIG α) (distance : Nat) (input : AIG.RefVec aig w)
+    (curr : Nat) (hcurr : curr ≤ w) (s : AIG.RefVec aig curr) :
+    ∀ (idx : Nat) (h1) (h2),
+        (go aig input distance curr hcurr s).aig.decls[idx]'h2 = aig.decls[idx]'h1 := by
+  generalize hgo : go aig input distance curr hcurr s = res
+  unfold go at hgo
+  split at hgo
+  . dsimp only at hgo
+    split at hgo
+    . rw [← hgo]
+      intro idx h1 h2
+      rw [blastShiftLeftConst.go_decl_eq]
+      rw [AIG.LawfulOperator.decl_eq (f := AIG.mkConstCached)]
+      apply AIG.LawfulOperator.lt_size_of_lt_aig_size (f := AIG.mkConstCached)
+      assumption
+    . rw [← hgo]
+      intro idx h1 h2
+      rw [blastShiftLeftConst.go_decl_eq]
+  . simp [← hgo]
+termination_by w - curr
+
+instance : AIG.LawfulVecOperator α AIG.ShiftTarget blastShiftLeftConst where
+  le_size := by
+    intros
+    unfold blastShiftLeftConst
+    apply blastShiftLeftConst.go_le_size
+  decl_eq := by
+    intros
+    unfold blastShiftLeftConst
+    apply blastShiftLeftConst.go_decl_eq
+
+namespace blastShiftLeft
+
+
+structure TwoPowShiftTarget (aig : AIG α) (w : Nat) where
+  n : Nat
+  lhs : AIG.RefVec aig w
+  rhs : AIG.RefVec aig n
+  pow : Nat
+
+def twoPowShift (aig : AIG α) (target : TwoPowShiftTarget aig w) : AIG.RefVecEntry α w :=
+  let ⟨n, lhs, rhs, pow⟩ := target
+  if h : pow < n then
+    let res := blastShiftLeftConst aig ⟨lhs, (2 ^ pow) % 2^n⟩
```

Comment:
I don't understand what's happening here. Isn't `(2^pow)%(2^n)` always equal to `2^pow`?